pkg/crds/crds.go

```diff
@@ -0,0 +1,160 @@
+				"confirmation": "yes-really-destroy-data",
+			},
+		},
+	}
+)
+
+func DeleteCustomResources(ctx context.Context, clientsets k8sutil.ClientsetsInterface, clusterNamespace string) {
+	err := deleteCustomResources(ctx, clientsets, clusterNamespace)
+	if err != nil {
+		logging.Fatal(err)
+	}
+	logging.Info("done")
+}
+
+func deleteCustomResources(ctx context.Context, clientsets k8sutil.ClientsetsInterface, clusterNamespace string) error {
+	for _, resource := range cephResources {
+		logging.Info(fmt.Sprintf("Getting resources kind %s", resource))
+		items, err := clientsets.ListResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, clusterNamespace)
+		if err != nil {
+			if strings.Contains(err.Error(), errorResourceNotFound.Error()) {
+				logging.Warning("the server could not find the requested resource: %s", resource)
+				continue
+			}
+			// log instead of return
+			return err
+		}
+
+		if len(items) == 0 {
+			logging.Warning("resource %s was not found on the cluster", resource)
+			continue
+		}
+
+		for _, item := range items {
+			logging.Info(fmt.Sprintf("removing resource %s: %s", resource, item.GetName()))
+			err = clientsets.DeleteResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, clusterNamespace, item.GetName())
+			if err != nil {
+				if err.Error() == "the server could not find the requested resource" {
+					logging.Info(err.Error())
+					continue
+				}
+				// log instead of return
+				return err
+			}
+
+			itemResource, err := clientsets.GetResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, item.GetName(), clusterNamespace)
+			if err != nil {
+				if err.Error() == "the server could not find the requested resource" {
+					logging.Info(err.Error())
+					continue
+				}
+				// log instead of return
+				return err
+			}
+
+			if itemResource != nil {
+				logging.Info("resource still alive, applying patch...")
+				err = updatingFinalizers(ctx, clientsets, itemResource, resource, clusterNamespace)
+				if err != nil {
+					if err.Error() == "the server could not find the requested resource" {
+						logging.Info(err.Error())
+						continue
+					}
+					// log instead of return
+					return err
+				}
+
+				err = clientsets.DeleteResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, clusterNamespace, item.GetName())
+				if err != nil {
+					if err.Error() == "the server could not find the requested resource" {
+						logging.Info(err.Error())
+						continue
+					}
+					// log instead of return
+					return err
+				}
+			}
+			logging.Info(fmt.Sprintf("resource %s was removed", item.GetName()))
+		}
+	}
+	return nil
+}
+
+func updatingFinalizers(ctx context.Context, clientsets k8sutil.ClientsetsInterface, itemResource *unstructured.Unstructured, resource, clusterNamespace string) error {
+	patch := map[string]interface{}{}
+
+	if itemResource.GetKind() == CephResourceKindCephCluster {
```

Comment:
If this is not a cephcluster resource, is any other patch applied? This seems like a no-op for other resources.

Reply:
Yes I'm adding an empty patch in the line 148, which is the default, this is changed only if the resource is of kind **CephCluster**

```go
patch := map[string]interface{}{}
```

Reply:
yes I added a default patch which is used for remove finalizers:

```go
const (
	defaultResourceRemoveFinalizers = map[string]interface{}{
		"metadata": map[string]interface{}{
			"finalizers": nil,
		},
	}
)
```